diff -ru ffmpeg-0.7.17/Makefile ../ffmpeg-0.7.17/Makefile
--- ffmpeg-0.7.17/Makefile	2015-03-12 01:05:23.000000000 +0100
+++ ../ffmpeg-0.7.17/Makefile	2022-01-08 18:07:39.193790500 +0100
@@ -120,6 +120,8 @@
 
 TEXIDEP = awk '/^@include/ { printf "$@: $(@D)/%s\n", $$2 }' <$< >$(@:%=%.d)
 
+POD2MAN = pod2man
+
 doc/%.html: TAG = HTML
 doc/%.html: doc/%.texi $(SRC_PATH_BARE)/doc/t2h.init
 	$(Q)$(TEXIDEP)
@@ -132,7 +134,7 @@
 
 doc/%.1: TAG = MAN
 doc/%.1: doc/%.pod
-	$(M)pod2man --section=1 --center=" " --release=" " $< > $@
+	$(M)$(POD2MAN) --section=1 --center=" " --release=" " $< > $@
 
 ifdef PROGS
 install: install-progs install-data
diff -ru ffmpeg-0.7.17/configure ../ffmpeg-0.7.17/configure
--- ffmpeg-0.7.17/configure	2015-03-12 01:05:42.000000000 +0100
+++ ../ffmpeg-0.7.17/configure	2022-01-08 18:07:39.162542900 +0100
@@ -158,6 +158,7 @@
   --list-filters           show all available filters
 
 External library support:
+  --enable-sdl             enable sdl [autodetect]
   --enable-avisynth        enable reading of AVISynth script files [no]
   --enable-bzlib           enable bzlib [autodetect]
   --enable-libcelt         enable CELT/Opus decoding via libcelt [no]
@@ -966,6 +967,7 @@
     rdft
     rtpdec
     runtime_cpudetect
+    sdl
     shared
     sinewin
     small
@@ -1243,6 +1245,8 @@
 
 symver_if_any="symver_asm_label symver_gnu_asm"
 
+speed_cflags=""
+
 # subsystems
 dct_select="rdft"
 mdct_select="fft"
@@ -1805,6 +1809,8 @@
     ;;
     --enable-debug=*) debuglevel="$optval"
     ;;
+    --enable-optimizations=*) speed_cflags="$optval"
+    ;;
     --disable-everything)
     map 'eval unset \${$(toupper ${v%s})_LIST}' $COMPONENT_LIST
     ;;
@@ -1950,7 +1956,7 @@
     cc_ident="llvm-gcc $($cc -dumpversion) $gcc_extra_ver"
     CC_DEPFLAGS='-MMD -MF $(@:.o=.d) -MT $@'
     AS_DEPFLAGS='-MMD -MF $(@:.o=.d) -MT $@'
-    speed_cflags='-O3'
+    test -z "${speed_cflags}" && speed_cflags='-O3'
     size_cflags='-Os'
 elif $cc -v 2>&1 | grep -qi ^gcc; then
     cc_type=gcc
@@ -1964,7 +1970,7 @@
         CC_DEPFLAGS='-MMD -MF $(@:.o=.d) -MT $@'
         AS_DEPFLAGS='-MMD -MF $(@:.o=.d) -MT $@'
     fi
-    speed_cflags='-O3'
+    test -z "${speed_cflags}" && speed_cflags='-O3'
     size_cflags='-Os'
 elif $cc --version 2>/dev/null | grep -q Intel; then
     cc_type=icc
@@ -1973,14 +1979,14 @@
     icc_version=$($cc -dumpversion)
     CC_DEPFLAGS='-MMD'
     AS_DEPFLAGS='-MMD'
-    speed_cflags='-O3'
+    test -z "${speed_cflags}" && speed_cflags='-O3'
     size_cflags='-Os'
     noopt_cflags='-O1'
 elif $cc -v 2>&1 | grep -q xlc; then
     cc_type=xlc
     cc_version="AV_STRINGIFY(__IBMC__)"
     cc_ident=$($cc -qversion 2>/dev/null | head -n1)
-    speed_cflags='-O5'
+    test -z "${speed_cflags}" && speed_cflags='-O5'
     size_cflags='-O5 -qcompact'
 elif $cc -V 2>/dev/null | grep -q Compaq; then
     cc_type=ccc
@@ -1989,7 +1995,7 @@
     DEPFLAGS='$(CPPFLAGS) $(CFLAGS) -M'
     debuglevel=3
     add_ldflags -Wl,-z,now # calls to libots crash without this
-    speed_cflags='-fast'
+    test -z "${speed_cflags}" && speed_cflags='-fast'
     size_cflags='-O1'
 elif $cc --vsn 2>/dev/null | grep -q "ARM C/C++ Compiler"; then
     test -d "$sysroot" || die "No valid sysroot specified."
@@ -2007,7 +2013,7 @@
     as_default="${cross_prefix}gcc"
     CC_DEPFLAGS='-MMD'
     AS_DEPFLAGS='-MMD'
-    speed_cflags='-O3'
+    test -z "${speed_cflags}" && speed_cflags='-O3'
     size_cflags='-Os'
     filter_asflags="filter_out -W${armcc_opt}*"
 elif $cc -version 2>/dev/null | grep -q TMS470; then
@@ -2023,7 +2029,7 @@
     add_cflags -D__gnuc_va_list=va_list -D__USER_LABEL_PREFIX__=
     CC_DEPFLAGS='-ppa -ppd=$(@:.o=.d)'
     AS_DEPFLAGS='-MMD'
-    speed_cflags='-O3 -mf=5'
+    test -z "${speed_cflags}" && speed_cflags='-O3 -mf=5'
     size_cflags='-O3 -mf=2'
     filter_cflags=tms470_flags
     tms470_flags(){
@@ -2057,7 +2063,7 @@
     cc_ident=$($cc --version | head -n1)
     CC_DEPFLAGS='-MMD'
     AS_DEPFLAGS='-MMD'
-    speed_cflags='-O3'
+    test -z "${speed_cflags}" && speed_cflags='-O3'
     size_cflags='-Os'
 elif $cc -V 2>&1 | grep -q Sun; then
     cc_type=suncc
@@ -2065,7 +2071,7 @@
     cc_ident=$($cc -V 2>&1 | head -n1 | cut -d' ' -f 2-)
     DEPEND_CMD='$(DEPCC) $(DEPFLAGS) $< | sed -e "1s,^.*: ,$@: ," -e "\$$!s,\$$, \\\," -e "1!s,^.*: , ," > $(@:.o=.d)'
     DEPFLAGS='$(CPPFLAGS) $(CFLAGS) -xM1'
-    speed_cflags='-O5'
+    test -z "${speed_cflags}" && speed_cflags='-O5'
     size_cflags='-O5 -xspace'
     filter_cflags=suncc_flags
     suncc_flags(){
@@ -2107,7 +2113,7 @@
     cc_ident=$($cc -v 2>&1 | head -n1 | tr -d :)
     CC_DEPFLAGS='-MMD -MF $(@:.o=.d) -MT $@'
     AS_DEPFLAGS='-MMD -MF $(@:.o=.d) -MT $@'
-    speed_cflags='-O2'
+    test -z "${speed_cflags}" && speed_cflags='-O2'
     size_cflags='-Os'
     filter_cflags='filter_out -Wdisabled-optimization'
 elif $cc -v 2>&1 | grep -q Open64; then
@@ -2116,7 +2122,7 @@
     cc_ident=$($cc -v 2>&1 | head -n1 | tr -d :)
     CC_DEPFLAGS='-MMD -MF $(@:.o=.d) -MT $@'
     AS_DEPFLAGS='-MMD -MF $(@:.o=.d) -MT $@'
-    speed_cflags='-O2'
+    test -z "${speed_cflags}" && speed_cflags='-O2'
     size_cflags='-Os'
     filter_cflags='filter_out -Wdisabled-optimization|-Wtype-limits|-fno-signed-zeros'
 fi
@@ -2960,6 +2966,7 @@
 enabled libxvid    && require  libxvid xvid.h xvid_global -lxvidcore
 enabled mlib       && require  mediaLib mlib_types.h mlib_VectorSub_S16_U8_Mod -lmlib
 
+if enabled sdl; then
 SDL_CONFIG="${cross_prefix}sdl-config"
 if check_pkg_config sdl SDL_version.h SDL_Linked_Version; then
     check_cpp_condition SDL.h "(SDL_MAJOR_VERSION<<16 | SDL_MINOR_VERSION<<8 | SDL_PATCHLEVEL) >= 0x010201" $sdl_cflags &&
@@ -2975,7 +2982,8 @@
     check_struct SDL.h SDL_VideoInfo current_w $sdl_cflags && enable sdl_video_size
   fi
 fi
-enabled sdl && add_cflags $sdl_cflags && add_extralibs $sdl_libs
+enabled sdl && enabled sdl_outdev && add_cflags $sdl_cflags && add_extralibs $sdl_libs
+fi
 
 texi2html -version > /dev/null 2>&1 && enable texi2html || disable texi2html
 
@@ -3261,6 +3269,7 @@
 echo "libxvid enabled           ${libxvid-no}"
 echo "zlib enabled              ${zlib-no}"
 echo "bzlib enabled             ${bzlib-no}"
+echo "sdl enabled               ${sdl-no}"
 echo
 
 for type in decoder encoder hwaccel parser demuxer muxer protocol filter bsf indev outdev; do
diff -ru ffmpeg-0.7.17/libavcodec/dxva2_internal.h ../ffmpeg-0.7.17/libavcodec/dxva2_internal.h
--- ffmpeg-0.7.17/libavcodec/dxva2_internal.h	2015-03-12 01:05:24.000000000 +0100
+++ ../ffmpeg-0.7.17/libavcodec/dxva2_internal.h	2022-01-08 19:24:14.253629700 +0100
@@ -28,6 +28,8 @@
 
 #include "config.h"
 
+#include "d3d9.h"
+
 #include "dxva2.h"
 #if HAVE_DXVA_H
 #include <dxva.h>
diff -ru ffmpeg-0.7.17/libavdevice/avdevice.h ../ffmpeg-0.7.17/libavdevice/avdevice.h
--- ffmpeg-0.7.17/libavdevice/avdevice.h	2015-03-12 01:05:26.000000000 +0100
+++ ../ffmpeg-0.7.17/libavdevice/avdevice.h	2022-01-08 18:40:20.538767800 +0100
@@ -59,5 +59,7 @@
  */
 void avdevice_register_all(void);
 
+//#undef strncpy
+
 #endif /* AVDEVICE_AVDEVICE_H */
 
diff -ru ffmpeg-0.7.17/libavdevice/dshow.h ../ffmpeg-0.7.17/libavdevice/dshow.h
--- ffmpeg-0.7.17/libavdevice/dshow.h	2015-03-12 01:05:26.000000000 +0100
+++ ../ffmpeg-0.7.17/libavdevice/dshow.h	2022-01-08 18:35:07.678362200 +0100
@@ -27,7 +27,8 @@
 #include <windows.h>
 #include <dshow.h>
 #include <dvdmedia.h>
-
+//#define wcscpy(a,b) StringCbCopyW(a,b)
+#undef wcscpy
 long ff_copy_dshow_media_type(AM_MEDIA_TYPE *dst, const AM_MEDIA_TYPE *src);
 void ff_print_AM_MEDIA_TYPE(const AM_MEDIA_TYPE *type);
 void ff_printGUID(const GUID *g);
diff -ru ffmpeg-0.7.17/libavformat/network.h ../ffmpeg-0.7.17/libavformat/network.h
--- ffmpeg-0.7.17/libavformat/network.h	2015-03-12 01:05:26.000000000 +0100
+++ ../ffmpeg-0.7.17/libavformat/network.h	2022-01-08 19:06:04.981030200 +0100
@@ -28,9 +28,11 @@
 #include "os_support.h"
 
 #if HAVE_WINSOCK2_H
+#define pollfd wsapollfd 
 #include <winsock2.h>
 #include <ws2tcpip.h>
-
+#undef pollfd
+//#define poll WSAPoll
 #define EPROTONOSUPPORT WSAEPROTONOSUPPORT
 #define ETIMEDOUT       WSAETIMEDOUT
 #define ECONNREFUSED    WSAECONNREFUSED
diff -ru ffmpeg-0.7.17/libavformat/os_support.c ../ffmpeg-0.7.17/libavformat/os_support.c
--- ffmpeg-0.7.17/libavformat/os_support.c	2015-03-12 01:05:26.000000000 +0100
+++ ../ffmpeg-0.7.17/libavformat/os_support.c	2022-01-08 19:07:08.528211300 +0100
@@ -61,7 +61,9 @@
 #if !HAVE_POLL_H
 #include <sys/time.h>
 #if HAVE_WINSOCK2_H
+#define pollfd wsapollfd
 #include <winsock2.h>
+#undef pollfd
 #elif HAVE_SYS_SELECT_H
 #include <sys/select.h>
 #endif
diff -ru ffmpeg-0.7.17/libavformat/os_support.h ../ffmpeg-0.7.17/libavformat/os_support.h
--- ffmpeg-0.7.17/libavformat/os_support.h	2015-03-12 01:05:26.000000000 +0100
+++ ../ffmpeg-0.7.17/libavformat/os_support.h	2022-01-08 19:07:51.496556900 +0100
@@ -60,7 +60,7 @@
 #define closesocket close
 #endif
 
-#if !HAVE_POLL_H
+#if !HAVE_POLL_H //&& !defined(_WIN32)
 typedef unsigned long nfds_t;
 
 struct pollfd {
diff -ru ffmpeg-0.7.17/libavutil/internal.h ../ffmpeg-0.7.17/libavutil/internal.h
--- ffmpeg-0.7.17/libavutil/internal.h	2015-03-12 01:05:26.000000000 +0100
+++ ../ffmpeg-0.7.17/libavutil/internal.h	2022-01-08 18:40:10.600042200 +0100
@@ -241,5 +241,5 @@
 #else /* HAVE_MMX */
 #define emms_c()
 #endif /* HAVE_MMX */
-
+#undef strncpy
 #endif /* AVUTIL_INTERNAL_H */
