--- src/wayland/fg_input_devices_wl.c.orig	2022-10-07 18:12:51.000000000 +0200
+++ src/wayland/fg_input_devices_wl.c	2023-09-27 13:29:54.626000000 +0200
@@ -139,7 +139,11 @@
     {
         fgSetWindow( window );
         xkb_keysym_to_utf8( sym, string, sizeof( string ) );
-        keyboard_cb( string[0], window->State.MouseX, window->State.MouseY, keyboard_ud );
+        for (unsigned int i = 0; string[i] && i < sizeof(string); ++i) {
+            keyboard_cb( string[i], window->State.MouseX, window->State.MouseY, keyboard_ud );
+             if ((string[ i ] & 0xC0) != 0xC0)
+                break ;
+        }
     }
 }
 
--- src/x11/fg_main_x11.c.orig	2022-10-07 18:12:51.000000000 +0200
+++ src/x11/fg_main_x11.c	2023-09-27 13:29:54.626000000 +0200
@@ -952,10 +952,14 @@
                     {
                         fgSetWindow( window );
                         fgState.Modifiers = fgPlatformGetModifiers( event.xkey.state );
-                        keyboard_cb( asciiCode[ 0 ],
-                                     event.xkey.x, event.xkey.y,
-                                     keyboard_ud
-                        );
+                        for (unsigned int i = 0; i < len; ++i) {
+                            keyboard_cb( asciiCode[ i ],
+                                         event.xkey.x, event.xkey.y,
+                                         keyboard_ud
+                            );
+                            if ((asciiCode[ i ] & 0xC0) != 0xC0)
+                                break ;
+                        }
                         fgState.Modifiers = INVALID_MODIFIERS;
                     }
                 }
